tests/ChunkedDecoderTest.php

```diff
@@ -0,0 +1,186 @@
+<?php
+
+namespace React\Tests\Http;
+
+use React\Stream\ReadableStream;
+use React\Http\ChunkedDecoder;
+
+class ChunkedDecoderTest extends TestCase
+{
+    public function setUp()
+    {
+        $this->input = new ReadableStream();
+        $this->parser = new ChunkedDecoder($this->input);
+    }
+
+    public function testSimpleChunk()
+    {
+        $this->parser->on('data', $this->expectCallableOnceWith('hello'));
+        $this->input->emit('data', array("5\r\nhello\r\n"));
+    }
+
+    public function testTwoChunks()
+    {
+    	$this->parser->on('data', $this->expectCallableConsecutive(2, array('hello', 'bla')));
+    	$this->input->emit('data', array("5\r\nhello\r\n3\r\nbla\r\n"));
+    }
+
+    public function testEnd()
+    {
+    	$this->parser->on('end', $this->expectCallableOnce(array()));
+    	$this->input->emit('data', array("0\r\n\r\n"));
+    }
+
+    public function testParameterWithEnd()
+    {
+    	$this->parser->on('data', $this->expectCallableConsecutive(2, array('hello', 'bla')));
+    	$this->parser->on('end', $this->expectCallableOnce(array()));
+    	$this->input->emit('data', array("5\r\nhello\r\n3\r\nbla\r\n0\r\n\r\n"));
+    }
+
+    public function testInvalidChunk()
+    {
+    	$this->parser->on('data', $this->expectCallableNever());
+    	$this->parser->on('error', $this->expectCallableOnce(array()));
+    	$this->input->emit('data', array("bla\r\n"));
+    }
+
+    public function testNeverEnd()
+    {
+    	$this->parser->on('end', $this->expectCallableNever());
+    	$this->input->emit('data', array("0\r\n"));
+    }
+
+    public function testWrongChunkHex()
+    {
+    	$this->parser->on('error', $this->expectCallableOnce(array()));
+    	$this->input->emit('data', array("2\r\na\r\n5\r\nhello\r\n"));
+    }
+
+    public function testSplittedChunk()
+    {
+        $this->parser->on('data',  $this->expectCallableOnceWith('welt'));
+
+        $this->input->emit('data', array("4\r\n"));
+        $this->input->emit('data', array("welt\r\n"));
+    }
+
+    public function testSplittedHeader()
+    {
+    	$this->parser->on('data',  $this->expectCallableOnceWith('welt'));
+
+    	$this->input->emit('data', array("4"));
+    	$this->input->emit('data', array("\r\nwelt\r\n"));
+    }
+
+    public function testSplittedBoth()
+    {
+    	$this->parser->on('data',  $this->expectCallableOnceWith('welt'));
+
+    	$this->input->emit('data', array("4"));
+    	$this->input->emit('data', array("\r\n"));
+    	$this->input->emit('data', array("welt\r\n"));
+    }
+
+    public function testCompletlySplitted()
+    {
+    	$this->parser->on('data',  $this->expectCallableOnceWith('welt'));
```

Comment:
This seems to indicate the whole chunk body is being buffered here? What if the chunk header says the chunk body will be several GiB?